pyeudiw/jwt/__init__.py

```diff
@@ -182,55 +220,73 @@ def sign(
         :returns: A string that represents the JWS.
         :rtype: str
         """
-        _key = key_from_jwk_dict(self.jwk.as_dict())
-
-        _payload: str | int | bytes = ""
 
+        _payload: str | int | bytes = plain_dict
+        _jwk = self.get_jwk_by_kid(kid) or self.jwks[0]
+        
         if isinstance(plain_dict, dict):
-            _payload = json.dumps(plain_dict).encode()
-        elif not plain_dict:
-            _payload = ""
+            _payload = json.dumps(plain_dict)
         elif isinstance(plain_dict, (str, int)):
             _payload = plain_dict
         else:
             _payload = ""
-        _signer = JWSec(_payload, alg=self.alg, **kwargs)
-        return _signer.sign_compact([_key], protected=protected, **kwargs)
 
-    def verify(self, jws: str, **kwargs) -> (str | Any | bytes):
+        _alg = DEFAULT_SIG_KTY_MAP[_jwk.kty]
+        _signer = JWSec(_payload, kty = _jwk.kty, alg=_alg, **kwargs)
+
+        if serialization_format=='compact':
+            return _signer.sign_compact(self.jwks, protected=protected)
```

Comment:
Passare tutto il set `self.jwks` potrebbe portare a dei bug.
In riga 225 si selezione una chiave di firma `_jwk`. Poi quella chiave non viene mai usata e si lascia alla libreria selezionare in qualche modo la chiave. In particolare, la libreria prende la prima chiave con compatibile con `alg` riportato nel token header, quindi ad es se ho multiple chiavi ECDSA su P-256, non ho garanzia di usare la chiave di verifica giusta.

Per una soluzione si può prendere supnto dai commit d0f27fa e 1108639 .

Reply:
mi sembra di capire che questo argomento consenta di specificare la chiave tra le tante caricate nel costruttore

in questo caso ha senso

Reply:
Non è esattamente così.

Passare self.jwks fa si che sia la libreria `cryptojwt` a scegliere quale chiave usare tra quelle caricate nel costruttore.
Questo è in totale antitesi con il resto della funzione.
In questa riga di codice
https://github.com/italia/eudi-wallet-it-python/blob/ab67361f149ad634478b316e579c8157db007365/pyeudiw/jwt/__init__.py#L225
si seleziona quale chiave usare per firmare.
Poi, invece che usare ***proprio*** quella chiave per firmare, speriamo che cryptojwk lo faccia veramente sulla base del kty e dell'alg. Se ho multiple chiavi nel self.jwks che soddisfano lo stesso kty e alg, non ho garanzia che cryptojwt selezioni la chiave giusta identificata nella riga di codice sopra.

Se noi abbiamo della logica nostra custom con la quale si identifica quale chiave usare per firmare, allora dovremmo dire a `cryptojwt` di firmare con la chiave che abbiamo identificato.


Questo discorso è totalmente analogo per i processi di verifica tokens, dove prima selezioniamo la chiave con cui verificare
https://github.com/italia/eudi-wallet-it-python/blob/ab67361f149ad634478b316e579c8157db007365/pyeudiw/jwt/__init__.py#L265-L287
ma poi non siamo sicuri di usarla: speriamo che cryptojwt la usi sulla base di euristiche sue
https://github.com/italia/eudi-wallet-it-python/blob/ab67361f149ad634478b316e579c8157db007365/pyeudiw/jwt/__init__.py#L291